### PR TITLE
Alerts page including lang, api and testing

### PR DESCRIPTION
**Pull request by laf**
Please see our [contributing docs](https://github.com/librenms/librenmsv2/blob/develop/docs/General/Contributing.md)
- [x] Have you created any necessary docs?
- [x] Have you created tests for Travis integration?
- [x] Have you used localisation in the WebUI?
- [x] Have you included the proper copyright?

Needs to have the acknowledge button adding to it but should be complete aside from that.

### CODE REVIEW

File: app/Models/Dashboard.php [hunk 1]
```diff
@@ -72,9 +72,9 @@ public function widgets()
      * @param $user_id
      * @return Builder|static
      */
-    public function scopeAllAvailable(Builder $query, $user_id)
```
Comment (murrant) [inline app/Models/Dashboard.php:75]: Oops, these changes snuck in.  can you revert them to usptream.

File: app/Api/Controllers/Alerting/AlertsController.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
+<?php
+/*
+ * Copyright (C) 2016 Neil Lathwood <neil@lathwood.co.uk>
```
Comment (murrant) [inline app/Api/Controllers/Alerting/AlertsController.php]: Your copyright/phpdoc header isn't matching the template here: https://github.com/librenms/librenmsv2/blob/develop/docs/Developing/File-Header.md

File: app/Api/Transformers/AlertsTransformer.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Turn this item object into a generic array
+     *
+     * @param Notification $alerts
+     * @return array
+     */
+    public function transform(Alerts $alerts)
+    {
+        return [
+            'id'        => (int)$alerts->id,
```
Comment (murrant) [inline app/Api/Transformers/AlertsTransformer.php]: Should more of these be cast to int?

File: app/DataTables/AlertsDataTable.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+     *
+     * @return array
+     */
+    protected function getBuilderParameters()
+    {
+        return [
+            'dom' => 'Blfrtip',
+            'lengthMenu' => [[25, 50, 100, -1], [25, 50, 100, "All"]],
+            'buttons' => [
+                'csv', 'excel', 'pdf', 'print', 'reset', 'reload',
```
Comment (murrant) [inline app/DataTables/AlertsDataTable.php:132]: Do we want just csv for now?

File: app/Models/Alerting/Rules.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+namespace App\Models\Alerting;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Builder;
+
+
+/**
+ *
+ */
+class Rules extends Model
```
Comment (murrant) [inline app/Models/Alerting/Rules.php]: The class name should be singular "Rule"

File: app/Models/Alerting/Rules.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    protected $table = 'alert_rules';
+    /**
+     * The primary key column name.
+     *
+     * @var string
+     */
+    protected $primaryKey = 'id';
+
+    // ---- Define Reletionships ----
```
Comment (murrant) [inline app/Models/Alerting/Rules.php]: Missing relationships

```
belongsTo('App\Models\Device')
hasMany('App\Models\Alerting\Alert')
```

Also, define the inverse of the relationship.

File: app/Models/Alerting/Alerts.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ 
+namespace App\Models\Alerting;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Builder;
+
+/**
+ *
+ */
+class Alerts extends Model
```
Comment (murrant) [inline app/Models/Alerting/Alerts.php]: Should be singular "Alert"

File: app/Models/Alerting/Alerts.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        return $this->belongsTo('App\Models\Alerting\Rules', 'rule_id');
+    }
+
+    /**
+     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
+     */
+    public function user()
+    {
+        return $this->belongsToMany('App\Models\User', 'devices_perms', 'device_id', 'user_id');
```
Comment (murrant) [inline app/Models/Alerting/Alerts.php]: Not sure this is needed as it is just an alias for `$alert->device->users` or with lazy eager loading `$alert->load('device.users;)`

If you haven't fetch the parent model yet, you can `Alert::active()->with('device.users')` for example

File: app/Models/Dashboard.php (hunk 1 above)
Comment (laf) [inline app/Models/Dashboard.php:75]: Should have mentioned it, I know it shouldn't be part of this pr but the tests break without this change as the original code was wrong.

File: app/Api/Controllers/Alerting/AlertsController.php (hunk 2 above)
Comment (laf) [inline app/Api/Controllers/Alerting/AlertsController.php]: I copied your copyright from one of the existing files, will update them all

File: app/DataTables/AlertsDataTable.php (hunk 4 above)
Comment (laf) [inline app/DataTables/AlertsDataTable.php:132]: They all work (for me at least), just the formatting on pdf one. I think leaving it will force us to fix it!

File: app/Models/Alerting/Rules.php (hunk 6 above)
Comment (laf) [inline app/Models/Alerting/Rules.php]: I was only defining them when needed, is it best practice to just add them?

Comment (laf) [inline app/Models/Alerting/Rules.php]: Also rules don't always belong to a device, they can be -1

File: app/Models/Alerting/Alerts.php (hunk 8 above)
Comment (laf) [inline app/Models/Alerting/Alerts.php]: This is used in App\Datatables\AlertsDataTable, I've not got it working any other way yet.

**Comment by laf** [comment]
Hopefully caught everything now, thanks @murrant

File: app/DataTables/AlertsDataTable.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+            ->editColumn('device.hostname', function ($this) {
+                return '<a href="' . url("devices/".$this['device']['device_id']) . '">' . $this['device']['hostname'] . '</a>';
+            })
+            ->addColumn('actions', function ($this) {
+                return '';
+            })
+            ->make(true);
+            //            ->addColumn('action', 'path.to.action.view')
+    }
+
```
Comment (murrant) [inline app/DataTables/AlertsDataTable.php:62]: I think this could be Alert::active()->with('device', 'user', rule')

File: app/DataTables/AlertsDataTable.php (hunk 4 above)
Comment (murrant) [inline app/DataTables/AlertsDataTable.php:132]: The thought I had was to reduce clutter.  I'm not sure how much you gain by exporting excel over csv

File: app/Models/Dashboard.php (hunk 1 above)
Comment (murrant) [inline app/Models/Dashboard.php:75]: Reason for this change was thinking we should in general prefer passing models over ids.

**Comment by murrant** [comment]
Missing the relationships on the Device Model to get rules and alerts.
Also, a decent amount of scrutinizer issues for you ;)

File: app/Models/Dashboard.php (hunk 1 above)
Comment (laf) [inline app/Models/Dashboard.php:75]: It was an id before, got an example of what you mean?

File: app/DataTables/AlertsDataTable.php (hunk 4 above)
Comment (laf) [inline app/DataTables/AlertsDataTable.php:132]: I don't think it's too bad, one extra button won't be much different.

File: app/DataTables/AlertsDataTable.php (hunk 9 above)
Comment (laf) [inline app/DataTables/AlertsDataTable.php:62]: Sorry I'm not sure what part of the code your referencing here.

**Comment by laf** [comment]
Hopefully got everything now.

I've chosen not to fix some scrut issues in things like the controllers where we are passing variables to functions that currently aren't used as they may get used later on. we can sweep through these in one go later on in the project.
